pkg/applicationserver/applicationserver.go

```diff
@@ -1352,6 +1403,15 @@ func (as *ApplicationServer) handleDownlinkNack(ctx context.Context, ids *ttnpb.
 			}
 
 			items := []*ttnpb.ApplicationDownlink{msg}
+
+			if msg.Confirmed && msg.ConfirmedRetry != nil {
+				msg.ConfirmedRetry.Attempt++
+				if msg.ConfirmedRetry.Attempt > msg.ConfirmedRetry.MaxAttempts.GetValue() {
```

Comment:
I would build the max attempts dynamically here:
1. If `confirmed_retry.max_attempts` is not null, validate it against the configured `max_attempts`.
2. Fallback to the default max attempts if it is not zero.
3. If both default and max are zero, do not drop.

I would also say that existing messages with `confirmed_retry == nil` should have their attempt counter initialized retroactively.